numpy/__init__.pyi

```diff
@@ -1,15 +1,52 @@
-# very simple, just enough to start running tests
-#
 import builtins
-from typing import Any, Mapping, List, Optional, Tuple, Union
+
+from typing import (
+    Any, Dict, Iterable, List, Optional, Mapping, Sized,
+    SupportsInt, SupportsFloat, SupportsComplex, SupportsBytes, SupportsAbs,
+    Tuple, Union,
+)
+
+import sys
 
 from numpy.core._internal import _ctypes
 
 _Shape = Tuple[int, ...]
 
+# Anything that can be coerced into numpy.dtype. To avoid recursive
+# definitions, any nested fields are required to be castable to a dtype object
+# are typed as Any.
+# Reference: https://docs.scipy.org/doc/numpy/reference/arrays.dtypes.html
+_DtypeLike = Union[
+    dtype,
+    # default data type
+    None,
+    # array-scalar types and generic types
+    type,  # TODO: enumerate these when we add type hints for numpy scalars
+    # TODO: add a protocol for anything with a dtype attribute
+    # character codes, type strings or comma-separated fields, e.g., 'float64'
+    str,
+    # (flexible_dtype, itemsize)
+    Tuple[Any, int],
+    # (fixed_dtype, shape)
+    Tuple[Any, _Shape],
```

Comment:
Note that `(int, 2)` is shorthand for `(int, (2,))`, so this the `int` case covers more than you might expect

Reply:
OK, in that case I'm going to use `_ShapeLike = Union[int, Tuple[int, ...]]` for clarity.